File: android/src/main/java/ch/nevis/mobile/authentication/sdk/react/MethodChannelHandler.kt
/*
 * Copyright © 2023 Nevis Security AG. All rights reserved.
 */

package ch.nevis.mobile.authentication.sdk.react

import ch.nevis.mobile.authentication.sdk.react.cache.OperationCache
import ch.nevis.mobile.authentication.sdk.react.model.ReactMethodName
import ch.nevis.mobile.authentication.sdk.react.model.message.outgoing.ChannelOutMessage
import ch.nevis.mobile.authentication.sdk.react.util.Json
import com.facebook.react.bridge.Promise
import kotlinx.serialization.encodeToString

private const val ERROR_CODE = "NMA_SDK"

object MethodChannelHandler {
    var channelData = HashMap<ReactMethodName, Promise>()

    fun resolve(method: ReactMethodName, message: ChannelOutMessage?) {
        channelData.remove(method)?.let { promise: Promise ->
            message?.let {
                val json = Json.instance.encodeToString(it)
                val map = Json.toMap(json)
                promise.resolve(map)
            } ?: promise.resolve(null)
        }
    }

    fun reject(method: ReactMethodName, message: ChannelOutMessage) {
        channelData.remove(method)?.let {
            val json = Json.instance.encodeToString(message)
            val map = Json.toMap(json)
            it.reject(ERROR_CODE, message.operationId, map)
        }

        OperationCache.removeOperation(message.operationId)
    }
}

File: ios/Handlers/MethodChannelHandler.swift
//
// Copyright © 2023 Nevis Security AG. All rights reserved.
//

struct MethodChannelData {
	let resolver: RCTPromiseResolveBlock
	let rejecter: RCTPromiseRejectBlock
}

class MethodChannelHandler {
	enum Constants: String {
		case errorCode = "NMA_SDK"
	}
	
	static let shared = MethodChannelHandler()

	// MARK: Properties

	var channelData: [ReactMethod: MethodChannelData] = [:]

	// MARK: Initialization

	private init() {}

	// MARK: Public Interface

	func resolve(method: ReactMethod, message: ChannelOutMessage?) {
		do {
			if let data = channelData.removeValue(forKey: method) {
				if let message {
					let arguments = try JsonMapper.encode(message)
					data.resolver(arguments)
				} else {
					data.resolver(nil)
				}
			}
		}
		catch {
			if let message {
				OperationCache.shared.delete(by: message.operationId)
			}
			print(error.localizedDescription)
		}
	}

	func reject(method: ReactMethod, message: ChannelOutMessage) {
		do {
			if let data = channelData.removeValue(forKey: method) {
				let arguments = try JsonMapper.encode(message)
				let error = NSError(domain: Constants.errorCode.rawValue, code: 0, userInfo: arguments)
				data.rejecter(Constants.errorCode.rawValue, message.operationId, error)
			}
		}
		catch {
			print(error.localizedDescription)
		}

		OperationCache.shared.delete(by: message.operationId)
	}
}
